Options for focus context presentation in braille

LeonarddeR opened the pull request:
### Link to issue number:
Fixes #217

### Summary of the issue:
from https://github.com/nvaccess/nvda/issues/217#issue-116021451

> When you arrow through a list of items and are reading the braille display, the start of each list item may begin in a different place for one list item than it is for another. This makes it more difficult to easily arrow down through a list in search of something, because you have to stop and read the display to be able to find the list item.

### Description of how this pull request fixes the issue:
This implements two types of focus context/ancestry presentation brought up by @JCSTeh:

1. https://github.com/nvaccess/nvda/issues/217#issuecomment-155279196: an option that allows you to always show the object's content at the very left of the braille display. This means that you will always have to scroll back if you want to see the ancestry, but this is no problem for speech oriented braille users and might even be their preference.
2. https://github.com/nvaccess/nvda/issues/217#issuecomment-303937777 : An option that only presents new focus ancestry information to the user when the ancestry has changed.

### Testing performed:
1. Go to braille settings
2. Set "Focus context presentation" to "Only when scrolling back", and press ok
3. Arrow around the desktop. Desktop items will show up at the left of the braille display
4. Set "Focus context presentation" to "Fill display for context changes", and press ok
5. When the desktop list is focused, the ancestry will be shown as earlier. When you start arrowing around, the selected desktop icon will be shown hard left, as there is no context change
6. Set "Focus context presentation" back to "Always fill display", and press ok
7. Arrowing through the desktop reveals default behaviour on the braille display.

### Known issues with pull request:
1. I'd like feedback from people about the wording of the new focus context presentation options and its options. It is quite hard to give a short description of the current, default behaviour actually. Also, I'm not that good in writing good summaries for the changes file.
2. This doesn't have a global command yet

### Open question
I'm using an integer in the config spec, is this preferred or should we go for a string instead?

Commit:
Added a focuspresentation option to the configSpec

Commit:
* When focusPresentation is set to 1, always show the focus object hard left
* Added focus presentation to the gui
* Updated copyright headers

dkager commented:
You'll also want to update the user guide once the final wording has been decided.

Commit:
First work on hybrid mode for braille focus presentation

Commit:
First attempt on getting the hybrid braille focus presentation to work, not yet ready for review

Commit:
Finished under de hood implementation of focus presentation hybrid mode, also did some rewording in the gui

Commit:
Reworded focus presentation setting in braille settings and changed the spec accordingly

Commit:
Reverted an unintentional line removal and fixed a translator comment

leonardder commented:
@JCSTeh: Not trying to push, but just wanted to let you know that this is ready for review.

One thing I'd like to point your attention at, is the use of a focusToHardLeftSet variable. I'm not particularly happy with this approach, but I can't come up with alternatives here.

dkager commented on source/braille.py:
```diff
@@ -1211,6 +1213,20 @@ def _set_windowEndPos(self, endPos):
 		if region.focusToHardLeft:
 			# Only scroll to the start of this region.
 			restrictPos = endPos - regionPos - 1
+		elif config.conf["braille"]["focusContextPresentation"]==2:
```
While NVDA certainly isn't too strict about this, I think having constants for this option would improve readability a lot.

leonardder replied:
If we'd go for constants, I'd say we should go for strings, or for 
option to be super modern. Items like tethering in the config spec 
should be option type as well, actually. Unless you can't set a default 
for those.

leonardder commented:
If we'd go for constants, I'd say we should go for strings, or for 
option to be super modern. Items like tethering in the config spec 
should be option type as well, actually. Unless you can't set a default 
for those.

leonardder commented on source/braille.py:
```diff
@@ -1211,6 +1213,20 @@ def _set_windowEndPos(self, endPos):
 		if region.focusToHardLeft:
 			# Only scroll to the start of this region.
```
This function has some double looping due to my recent changes, so I'm planning to change the order a bit later today. I want it to loop through the regions only once.

Commit:
For _set_windowEndPos, loop through the current currently visible regions only once

Commit:
Changed focus context presentation to an option based option in the config spec

Commit:
Provided a unit test for #217

leonardder commented:
I just added unit tests for this.

jcsteh commented on source/gui/settingsDialogs.py:
```diff
@@ -1623,6 +1623,30 @@ def makeSettings(self, settingsSizer):
+			# Translators: The label for a braille focus context presentation setting that
+			# shows as much as possible focus context information if the focus object doesn't fill up the whole display.
+			# This was the pre NVDA 2017.3 default.
+			_("Always fill display")),
```
nit: While I understand you're trying to get the translators comment as close to the translatable string as possible, this does make it a bit hard to read this list. Happy to discuss, but i think this looks more readable despite separating it a bit:

> ```
> # Translators: Blah blah blah
> ("fill", _("Always fill display")),
> ```

I'm pretty sure gettext will still match this - I'm pretty sure we do it elsewhere - but maybe I should check...

jcsteh commented on source/gui/settingsDialogs.py:
```diff
@@ -1623,6 +1623,30 @@ def makeSettings(self, settingsSizer):
+		focusContextPresentationChoices = [x[1] for x in self.focusContextPresentations]
+		self.focusContextPresentationList = sHelper.addLabeledControl(focusContextPresentationLabelText, wx.Choice, choices=focusContextPresentationChoices)
+		try:
+			index=[x[0] for x in self.focusContextPresentations].index(config.conf["braille"]["focusContextPresentation"])
```
Since we end up having to run this list comprehension again later, I think it'd make sense to save this on self. That would also make the code more readable in both places.

jcsteh commented on tests/unit/test_braille.py:
```diff
@@ -0,0 +1,96 @@
+		"""Set up a fake focus object and give it some ancestry."""
+		self.obj=NVDAObjectWithRole(role=controlTypes.ROLE_LISTITEM)
+		# Create and set a fake desktop object
+		self.fakeDesktopObj=NVDAObjectWithRole(role=controlTypes.ROLE_WINDOW)
```
Consider moving this into the bootstrap code, just as we set the focus and navigator objects.

jcsteh commented on tests/unit/test_braille.py:
```diff
@@ -0,0 +1,96 @@
+		self.assertEqual(len(self.regionsWithPositions),3)
+
+	def test_fillDisplay(self):
+		"""Test for the case where both the focus object and its ancestors should be visible on a 40 cell display."""
```
Perhaps "all its ancestors" to be super clear?

jcsteh commented on tests/unit/test_braille.py:
```diff
@@ -0,0 +1,96 @@
+		# This is because getting windowEndPos can update windowStartPos
+		# Both the focus object and its parent should fit on the display
+		# Thus, the window end position is equal to the end position of the 3rd region
+		self.assertEqual(braille.handler.buffer.windowEndPos,self.regionsWithPositions[2][2])
```
If it's not too much trouble, consider making the regions with positions code in braille return a namedtuple. That would make these tests more readable; e.g. `self.regionsWithPositions[2].end`.

jcsteh commented on tests/unit/test_braille.py:
```diff
@@ -0,0 +1,96 @@
+		# Only the focus object should be visible now, equivalent to scroll only
+		self.assertEqual(braille.handler.buffer.windowEndPos,self.regionsWithPositions[2][2])
+		self.assertEqual(braille.handler.buffer.windowStartPos,self.regionsWithPositions[2][1])
+		# Clean up the cached focus ancestors
```
You're only cleaning up ancestors to a certain point, so I think this comment needs some clarification. You could use the roles to help explain.

jcsteh commented on tests/unit/test_braille.py:
```diff
@@ -0,0 +1,96 @@
+		# Forcefully create a fake focus ancestry
+		globalVars.focusAncestors=[api.getDesktopObject(),NVDAObjectWithRole(role=controlTypes.ROLE_DIALOG),NVDAObjectWithRole(role=controlTypes.ROLE_LIST)]
+		braille.handler.handleGainFocus(self.obj)
+		# Make sure that we are testing with three regions
```
It'd be super helpful if this comment could be expanded to list the roles (dialog, list, list item) and note that braille excludes the desktop object. I forgot about this last point myself and was thus very confused.

jcsteh commented on tests/unit/test_braille.py:
```diff
@@ -0,0 +1,96 @@
+		braille.handler.handleGainFocus(self.obj)
+		# WindowEndPos should be retrieved before we attempt to get the start position
+		# This is because getting windowEndPos can update windowStartPos
+		# Both the focus object and its parent should fit on the display
```
"its parent" -> "its ancestors"

jcsteh reviewed (changes requested):
This looks pretty good for the most part. Very nice work. Thanks.

jcsteh commented:
Oh, and this now has merge conflicts. :(

Commit:
Merge branch 'master' into i217, resolving merge conflict

leonardder commented on source/braille.py:
```diff
@@ -1398,18 +1406,23 @@ def getFocusContextRegions(obj, oldFocusRegions=None):
 		region = NVDAObjectRegion(parent, appendText=TEXT_SEPARATOR)
 		region._focusAncestorIndex = index
+		if config.conf["braille"]["focusContextPresentation"]=='hybrid' and not focusToHardLeftSet:
+			region.focusToHardLeft = True
```
> Would this be more readable if it just did  index == newAncestorsStart ? That would avoid the need for the  focusToHardLeftSet  flag and also make it clear that we do this only for the first new ancestor.

This is not possible due to the "if not parent.isPresentableFocusAncestor:" line.

leonardder commented on source/braille.py:
```diff
@@ -1398,18 +1406,23 @@ def getFocusContextRegions(obj, oldFocusRegions=None):
 			commonRegionsEnd = 0
 			newAncestorsStart = 1
-		# Yield the common regions.
+		# Yield the common regions, thereby setting focusToHardLeft back to False.
```
Ah, this was part of my initial implementation of this and I wanted to make sure that all cached regions with focusToHardLeft set to True had it set back to False. But, this isn't necessary.

leonardder replied:
Hmm, no, the current implementation still relies on setting focusTohardLeft to False. Let me investigate this in more detail and come up with a descriptive comment. Cheers for unit testing.

leonardder commented on source/braille.py:
```diff
@@ -1677,7 +1690,16 @@ def handleGainFocus(self, obj):
 
 	def _doNewObject(self, regions):
 		self.mainBuffer.clear()
+		focusToHardLeftSet = False
```
> Would the "functional" purpose of this be clearer if it were named, for example,  foundHardLeft ?

Honestly I don't think so, as we aren't only checking for focusToHardLeft on a region, but also setting it when the focus ancestry is equal.

leonardder commented on source/braille.py:
```diff
@@ -1677,7 +1690,16 @@ def handleGainFocus(self, obj):
+			if self.tether == self.TETHER_FOCUS and config.conf["braille"]["focusContextPresentation"]=='hybrid':
+				if region.focusToHardLeft:
+					focusToHardLeftSet = True
+				elif not focusToHardLeftSet and getattr(region, "_focusAncestorIndex", None) is None:
```
I'm afraid your incorrect here. This code applies to the first region that is returned by getFocusRegions, and this doesn't necessarily have to be regions-1 if I'm correct.

leonardder commented on source/config/configSpec.py:
```diff
@@ -62,6 +62,7 @@
 	tetherTo = string(default="focus")
 	readByParagraph = boolean(default=false)
 	wordWrap = boolean(default=true)
+	focusContextPresentation = option("fill","scroll","hybrid",default="fill")
```
If you want a reorder, do you also want a reorder in the gui and config spec?

Commit:
Review actions part 1

leonardder commented:
note that this not yet ready for another round of review, there are some 
open comments.

As for the review regarding the unit tests, I decided to create a new 
module called objectProvider which we can use to add more object related 
stuff to if we so desire later.

Commit:
Reoder focus context presentation options to show the new default first

Commit:
First attempt to come up with documentation for the new focus context presentation option. Still a bit rough though.

jcsteh commented on tests/unit/__init__.py:
```diff
@@ -66,15 +66,17 @@ class AppArgs:
 api.setFocusObject(phObj)
 api.setNavigatorObject(phObj)
-
+api.setDesktopObject(phObj)
```
nit: Reinstate blank line after this.

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
 
+==== Focus context presentation ====
+This option allows you to choose what context information NVDA will show on the braille display when an object gets focus.
+Context information (or focus ancestry) refers to the hierarchical tree of objects containing the focus object.
```
I'm concerned about terminology overload here. Perhaps simplify to something like:

> `Context information (or focus ancestry) refers to the hierarchy of objects containing the focus.`

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
+==== Focus context presentation ====
+This option allows you to choose what context information NVDA will show on the braille display when an object gets focus.
+Context information (or focus ancestry) refers to the hierarchical tree of objects containing the focus object.
+For example, when you just focused a list item, this list item is part of a list. This list might be contained by a dialog, etc.
```
1. "when you just focused" -> "when you focus"
2. Two sentences on one line. Please split.

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
+When set to fill display for context changes, NVDA will try to display as much as possible context information on the braille display, but only for the parts of the context that have changed.
+For the example above, this means that when changing focus to the list, NVDA will show the list item on the braille display.
+Furthermore, if there is enough space left on the braille display, NVDA will try to show that the list item is part of a list.
+Now, if you start moving through the list with your arrow keys, it is assumed that you are aware of the fact that you are moving through a list.
```
Perhaps this:

> `If you then start moving through the list with your arrow keys, it is assumed that you are aware that you are still in the list.`

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
+Thus, for the remaining list items you focus, NVDA will only show the focused list item on the display.
+In order for you to read the context again (i.e. that you are in a list and that the list is part of a dialog), you will have to scroll your braille display back.
+
+When this option is set to always fill the display, NVDA will try to show as much as possible context information on the braille display, regardless of whether you have seen the same context information before.
```
"as much as possible context information" -> "as much context information as possible"

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
+In order for you to read the context again (i.e. that you are in a list and that the list is part of a dialog), you will have to scroll your braille display back.
+
+When this option is set to always fill the display, NVDA will try to show as much as possible context information on the braille display, regardless of whether you have seen the same context information before.
+This has the advantage that NVDA will make use of the braille display in an effective way, filling up as much as possible empty display space.
```
I know what you mean here, but some users (you and Davy, for example, heh) might argue this isn't "effective". Perhaps this:

> `This has the advantage that NVDA will fit as much information as possible on the display.`

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
+
+When this option is set to always fill the display, NVDA will try to show as much as possible context information on the braille display, regardless of whether you have seen the same context information before.
+This has the advantage that NVDA will make use of the braille display in an effective way, filling up as much as possible empty display space.
+however, this option has the downside that there is always a difference in the position where the representation of the object with focus starts on the braille display.
```
1. Need capital at start of sentence.
2. I think we can simplify/shorten this:

> `However, the downside is that there is always a difference in the position where the focus starts on the braille display.`

Would it also be useful to mention that this makes skimming difficult? For example:

> `This can make it difficult to skim a long list of items, for example, as you will need to continually move your finger to find the start of the item.`

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,27 @@ When you scroll the display, you will then be able to read the rest of the word.
+This was the default behavior for NVDA 2017.2 and before.
+
+When you set the focus context presentation option to only show the context information when scrolling back, NVDA never shows context information on your braille display by default.
+Thus, in the example above, it is presented on the braille display that you focused a list item. However, in order for you to read the context (i.e. that you are in a list and that the list is part of a dialog), you will have to scroll your braille display back.
```
1. "it is presented on the braille display that" -> "NVDA will display that"
2. Two sentences on one line. Please split.

jcsteh reviewed (changes requested):
Very nice comments and documentation. Thank you. Almost there. :)

Commit:
Review actions part 2

Commit:
Add a script to globalCommands

leonardder commented:
[**@leonardder**](https://github.com/leonardder) commented on [11 jul. 2017 18:37 CEST](https://github.com/nvaccess/nvda/pull/7236#issuecomment-314501755 "2017-07-11T16:37:42Z - Replied by Github Reply Comments"):
> note that this not yet ready for another round of review

@JCSTeh: It seems you decided otherwise, many thanks. I've addressed your comments.

I also added a script to globalCommands, which I left unbound by default. I'd argue that users don't change this every single day. If you think this should be bound by default, I propose control+NVDA+F for focus.

Commit:
Fix accidental override of miscDeps

jcsteh commented on source/globalCommands.py:
```diff
@@ -1698,6 +1698,23 @@ def script_braille_toggleTether(self, gesture):
+			index = values.index(config.conf["braille"]["focusContextPresentation"])
+		except:
+			index=0
+		newIndex = index+1 if index+1<len(values) else 0
```
FYI, this can be shortened to:

> `newIndex = index % len(values)`

I'll leave it up to you as to whether you want to change this. Whether it makes things more readable or not is certainly debatable. :)

jcsteh commented on source/globalCommands.py:
```diff
@@ -1698,6 +1698,23 @@ def script_braille_toggleTether(self, gesture):
+		config.conf["braille"]["focusContextPresentation"] = values[newIndex]
+		braille.invalidateCachedFocusAncestors(0)
+		braille.handler.handleGainFocus(api.getFocusObject())
+		# Translators: Reports the new state of braille focus context presentation.
```
It might be helpful to note that %s gets replaced with the context presentation setting and give an example:

> ```
> # %s will be replaced with the context presentation setting.
> # For example, the full message might be "Braille focus context presentation Fill display for context changes"
> ```

Actually, given the length of that example, I wonder whether we need a colon here. Normally, we prefer to avoid this for brevity (linguistic correctness is less important for these kinds of things), but this is long enough that it might be confusing without it. Alternatively, we could just cut the prefix altogether and just use the setting name. Arguably, the setting names are long/unique enough that the context will be fairly clear.

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,32 @@ When you scroll the display, you will then be able to read the rest of the word.
 
+==== Focus context presentation ====
+This option allows you to choose what context information NVDA will show on the braille display when an object gets focus.
+Context information (or focus ancestry) refers to the hierarchy of objects containing the focus.
```
I still wonder whether we should drop the term "focus ancestry" here. I'm concerned it will just confuse most users and it's certainly more technical than "context information".

jcsteh commented on user_docs/en/userGuide.t2t:
```diff
@@ -1074,6 +1074,32 @@ When you scroll the display, you will then be able to read the rest of the word.
+This list might be contained by a dialog, etc.
+Please consult the section about [object navigation #ObjectNavigation] for more information about the hierarchy that applies to objects in NVDA.
+
+When set to fill display for context changes, NVDA will try to display as much as possible context information on the braille display, but only for the parts of the context that have changed.
```
"as much as possible context information" -> "as much context information as possible"

Commit:
Review actions part 3
